BJS2-15681 - Запрос достижений

Kostahe opened the pull request:
**Implement AchievementController and DTOs for Achievement Management**

- **Added `AchievementController`** with REST endpoints:
  - **`GET /achievements`**: Fetch achievements by filter criteria.
  - **`GET /users/{userId}`**: Retrieve achievements for a specific user by their ID.
  - **`GET /{id}`**: Get a specific achievement by its ID.
  - **`GET /users/{userId}/progress`**: Get progress on achievements for a specific user.

- **Introduced new DTO classes**:
  - **`AchievementDto`**: Represents detailed information about an achievement.
  - **`AchievementFilterDto`**: Contains filter criteria for querying achievements.
  - **`AchievementProgressDto`**: Represents progress information for an achievement.

- **Created `AchievementFilter` implementations**:
  - **`AchievementDescriptionFilter`**: Filters achievements based on their description.
  - **`AchievementRarityFilter`**: Filters achievements based on their rarity.
  - **`AchievementTitleFilter`**: Filters achievements based on their title.

- **Updated `AchievementService`**:
  - Implemented methods to interact with the repositories and apply filters as defined in the new `AchievementFilter` classes.

- **Enhanced error handling** with `EntityNotFoundException` and a global exception handler for standardized error responses.

Kostahe committed:
BJS2-15681 - Запрос достижений

**Implement AchievementController and DTOs for Achievement Management**

- **Added `AchievementController`** with REST endpoints:
  - **`GET /achievements`**: Fetch achievements by filter criteria.
  - **`GET /users/{userId}`**: Retrieve achievements for a specific user by their ID.
  - **`GET /{id}`**: Get a specific achievement by its ID.
  - **`GET /users/{userId}/progress`**: Get progress on achievements for a specific user.

- **Introduced new DTO classes**:
  - **`AchievementDto`**: Represents detailed information about an achievement.
  - **`AchievementFilterDto`**: Contains filter criteria for querying achievements.
  - **`AchievementProgressDto`**: Represents progress information for an achievement.

- **Created `AchievementFilter` implementations**:
  - **`AchievementDescriptionFilter`**: Filters achievements based on their description.
  - **`AchievementRarityFilter`**: Filters achievements based on their rarity.
  - **`AchievementTitleFilter`**: Filters achievements based on their title.
  - **`AchievementTitleFilter`**: Filters achievements based on their title.

- **Updated `AchievementService`**:
  - Implemented methods to interact with the repositories and apply filters as defined in the new `AchievementFilter` classes.

- **Enhanced error handling** with `EntityNotFoundException` and a global exception handler for standardized error responses.

GevorgSogomonian commented on src/main/java/faang/school/achievement/service/AchievementService.java:
```diff
@@ -0,0 +1,68 @@
+
+    @Transactional(readOnly = true)
+    public AchievementDto getAchievementById(long id) {
+        Achievement achievement = achievementRepository.findById(id)
+            .orElseThrow(() -> new EntityNotFoundException(String.format("Achievement with ID: %d not found", id)));
```
Почему ты только в этом методе кидаешь ошибку, если сущность не в бд?

Kostahe replied:
Да, это правда, я пропустил это. Спасибо!

GevorgSogomonian commented on src/main/java/faang/school/achievement/controller/AchievementController.java:
```diff
@@ -0,0 +1,38 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("achievements")
```
Здесь мне кажется ты забыл "/" перед "achivements"

Kostahe replied:
Проверь, как путь формируется в контексте всего контроллера, и от этого зависят правила. Обычно путь на уровне класса может быть как с начальным /, так и без него.

GevorgSogomonian commented on src/main/java/faang/school/achievement/dto/ErrorResponse.java:
```diff
@@ -0,0 +1,7 @@
+package faang.school.achievement.dto;
```
Мне кажется лучше в названии файла помечать что это dto, а то как-то странно выглядит, когда на структуру файлов смотришь, все файлы в папке помечены как dto, а этот нет, выглядит будто он здесь оказался по ошибке

Kostahe replied:
Да, я согласен. Спасибо за замечание!

GevorgSogomonian reviewed (commented):
Круто пишешь. Посмотри пару моментов по комментам

Kostahe committed:
BJS2-15681 - Запрос достижений

- **Added `AchievementController`** with REST endpoints:
  - **`GET /achievements`**: Fetch achievements by filter criteria.
  - **`GET /users`**: Retrieve achievements for a specific user by their ID.
  - **`GET /{id}`**: Get a specific achievement by its ID.
  - **`GET /users/progress`**: Get progress on achievements for a specific user.

- **Introduced new DTO classes**:
  - **`AchievementDto`**: Represents detailed information about an achievement.
  - **`AchievementFilterDto`**: Contains filter criteria for querying achievements.
  - **`AchievementProgressDto`**: Represents progress information for an achievement.

- **Created `AchievementFilter` implementations**:
  - **`AchievementDescriptionFilter`**: Filters achievements based on their description.
  - **`AchievementRarityFilter`**: Filters achievements based on their rarity.
  - **`AchievementTitleFilter`**: Filters achievements based on their title.

- **Updated `AchievementService`**:
  - Implemented methods to interact with the repositories and apply filters as defined in the new `AchievementFilter` classes.

- **Enhanced error handling** with `EntityNotFoundException` and a global exception handler for standardized error responses.

Kostahe committed:
BJS2-15681 - Запрос достижений

- removed getById from AchievementRepository

Kostahe committed:
Merge branch 'titan-master-bc5' into feature-BJS2-15681
